gateway/src/main/java/io/camunda/zeebe/gateway/impl/stream/ClientStreamAdapter.java

```diff
@@ -0,0 +1,98 @@
+
+import io.camunda.zeebe.gateway.ResponseMapper;
+import io.camunda.zeebe.gateway.protocol.GatewayOuterClass.ActivatedJob;
+import io.camunda.zeebe.gateway.protocol.GatewayOuterClass.StreamActivatedJobsRequest;
+import io.camunda.zeebe.msgpack.value.StringValue;
+import io.camunda.zeebe.protocol.impl.stream.job.ActivatedJobImpl;
+import io.camunda.zeebe.protocol.impl.stream.job.JobActivationProperties;
+import io.camunda.zeebe.protocol.impl.stream.job.JobActivationPropertiesImpl;
+import io.camunda.zeebe.scheduler.future.ActorFuture;
+import io.camunda.zeebe.transport.stream.api.ClientStreamConsumer;
+import io.camunda.zeebe.transport.stream.api.ClientStreamId;
+import io.camunda.zeebe.transport.stream.api.ClientStreamer;
+import io.grpc.stub.ServerCallStreamObserver;
+import java.util.concurrent.CompletableFuture;
+import org.agrona.DirectBuffer;
+
+public class ClientStreamAdapter {
+  private final ClientStreamer<JobActivationProperties> jobStreamer;
+
+  public ClientStreamAdapter(final ClientStreamer<JobActivationProperties> jobStreamer) {
+    this.jobStreamer = jobStreamer;
+  }
+
+  public void handle(
+      final StreamActivatedJobsRequest request,
+      final ServerCallStreamObserver<ActivatedJob> responseObserver) {
+    final JobActivationProperties jobActivationProperties = toJobActivationProperties(request);
+
+    final ActorFuture<ClientStreamId> clientStreamId =
+        jobStreamer.add(
+            wrapString(request.getType()),
+            jobActivationProperties,
+            new ClientStreamConsumerImpl(responseObserver));
+
+    final var removeJobStream = new JobStreamRemover(clientStreamId, jobStreamer);
+    responseObserver.setOnCloseHandler(removeJobStream);
+    responseObserver.setOnCancelHandler(removeJobStream);
+  }
+
+  private JobActivationProperties toJobActivationProperties(
+      final StreamActivatedJobsRequest request) {
+    final JobActivationPropertiesImpl jobActivationProperties = new JobActivationPropertiesImpl();
+    final DirectBuffer worker = wrapString(request.getWorker());
+    jobActivationProperties.setWorker(worker, 0, worker.capacity());
+    jobActivationProperties.setTimeout(request.getTimeout());
+    jobActivationProperties.setFetchVariables(
+        request.getFetchVariableList().stream().map(StringValue::new).toList());
+
+    return jobActivationProperties;
+  }
+
+  static class ClientStreamConsumerImpl implements ClientStreamConsumer {
+    private final ServerCallStreamObserver<ActivatedJob> responseObserver;
+
+    public ClientStreamConsumerImpl(final ServerCallStreamObserver<ActivatedJob> responseObserver) {
+      this.responseObserver = responseObserver;
+    }
+
+    @Override
+    public CompletableFuture<Void> push(final DirectBuffer payload) {
+      // TODO use gRPCExecutor thread pool
+      return CompletableFuture.runAsync(
+          () -> {
+            final ActivatedJobImpl deserializedJob = new ActivatedJobImpl();
+            deserializedJob.wrap(payload);
```

Comment:
:x: We should add a `TODO` here and link it to the GitHub issue for validating the payload against the `JobActivationProperties`.

In the kickoff, we agreed that we need to perform the validation, but we need to decide what happens to the job if the validation fails.

Reply:
I think, we can opt out validating the payload returned from the broker for now because it does not directly improve the UX but provides better visibility when broker returns unexpected job. At the same time, we can add the validation for user input where we prevent receiving invalid job activation properties, so prevent not returning any jobs to the worker. What do you think? @koevskinikola

Cc: @npepinpe

Reply:
It may improve UX in the future to be defensive here and ensure we're not streaming the wrong job to a client, but as we discussed on Slack, it can be postponed :shrug:

Reply:
All good from my side, we can skip it for now.